### ISSUE TITLE
Consider implementing a maximum density difference report

### ISSUE THREAD

**Issue by schmie** [schmie/hemeMove#58]
**Reported by miguel on 14 Oct 2011 13:11 UTC**
I'm currently developing a set of tests against analytic solutions of Poiseuille flow in 3D. I've noticed that in order to get good match, you need to make sure that the system remains in the quasi-incompressible regime (i.e. moderate density gradients).

This is nothing new and I assume it has to do with the error of order squared Mach number that appears in the derivation of NS from LB. For a fixed maximum velocity and space step, reducing the time step reduces Mach number and therefore your error.

A good example of this is the current configuration of difftest which does not give a good match with an analytic solution due to differences of density of up to 17%. Reducing time step by a factor of 4 reduces this difference down to 1.5% and gives pretty decent matches (modulo the non-zero velocity at the boundary due to bounce-back being use). See plot attached.

I would therefore suggest implementing a small check for maximum density difference at the end of the simulation that gives a hint to the user about how trustworthy the results generated are.

**Comment by schmie** [comment]
**Modified by miguel on 14 Oct 2011 13:11 UTC**

**Comment by schmie** [comment]
**Comment by hywel on 14 Oct 2011 14:21 UTC**
Compelling graph. I think that we'd ideally do it at some other point than the end of the simulation though - it seems like something you'd want to know as soon as possible.

Maybe instead, we could make it operate in the same way as the stability tester, but outputting a maximum density difference / density standard deviation instead of killing the simulation.

**Comment by schmie** [comment]
**Modified by hywel on 20 Oct 2011 17:25 UTC**

**Comment by schmie** [comment]
**Comment by miguel on 15 Nov 2011 16:42 UTC**
Hywel, can you please check my proposed design?

How about:
* Creating a new `IteratedAction` subclass called `AccuracyTester` (or `IncompressibilityTester`) that keeps track of the max/min value of rho at each subdomain.
* Create an instance of that class in `SimulationMaster` and add it to `actors`.
* Once the simulation is over (`SimulationMaster::PostSimulation`), let the IO process write to the report file the global max/min. 

In order to get global max/min:
* Is it possible (and sensible) to make `AccuracyTester` a `PhasedBroadcastIrregular` subclass that does a single pass up/down the tree on the very last time step to find out global max/min? 
* In that case, is it safe to assume that the IO proc will always be the tree root so that only a pass up is required?
* Alternately, can we tolerate the overhead of `MPI_Reduce`-ing twice at the master proc once the simulation is over? 

In order to unit test this functionality, we would need to decide on a strategy to test code in parallel and probably put together the infrastructure. Has it been discussed before? Otherwise, let's talk about it on Thursday.

**Comment by schmie** [comment]
**Modified by miguel on 15 Nov 2011 16:43 UTC**

**Comment by schmie** [comment]
**Comment by hywel on 15 Nov 2011 18:07 UTC**
I'd suggest `IncompressibilityChecker`, rather than `Tester`.

I'm not convinced it's something we only want at the very end of the simulation - if it has a significant effect on accuracy we should be reporting sooner than that to give a user a chance to kill the simulation and start again. The extra computation to do this will be tiny.

Consider also that if HemeLB were being used by neurosurgeons, they almost certainly wouldn't set a limit on the number of cycles in their simulation, they'd just let it run for as long as they needed it, hence would never see the message about poor accuracy.

I think it's safe to assume that the IO proc will always be the root of the tree, but feel free to add a comment to !PhasedBroadcast.h making clear what will fail if we one decide to change the tree root.

Yup, no idea how we reliably unit-test parallel bits of the code.

**Comment by schmie** [comment]
**Comment by miguel on 25 Nov 2011 15:22 UTC**
83844a098cf7 preliminary implementation of `IncompressibilityChecker` plus unit test.

James, please review. Specially the part to do with the broadcast mock. Cheers.

**Comment by schmie** [comment]
**Comment by jamespjh on 25 Nov 2011 17:47 UTC**
CR OK, comments given verbally.

**Comment by schmie** [comment]
**Comment by miguel on 30 Nov 2011 11:39 UTC**
Replying to [jamespjh](comment:8):
> CR OK, comments given verbally.

This involved:
 * Moving the broadcast mock to the unit tests folder (needs the `.hpp,.h,.cc` trick in `IncompressibilityChecker`).
 * Testing default largest and smallest densities against arithmetic expressions.

Done in 6a3e199986d4

**Comment by schmie** [comment]
**Comment by miguel on 5 Dec 2011 14:52 UTC**
Incompressibility checker preliminary integrated in `SimulationMaster` in a1d4c8e7acb0. 

Hywel, please CR. In particular the comment that I added to `PhasedBroadcast.h` which documents the feature that was causing me trouble last week.

We need to decide on a policy on how to react to incompressibility violations (at the moment, the checker is cluttering stdout with messages). Ideally something consistent with stability testing.

**Comment by schmie** [comment]
**Comment by miguel on 5 Dec 2011 14:55 UTC**
To do:
 * evaluate impact on performance of the checker.
 * several TODO comments left in the code.

**Comment by schmie** [comment]
**Comment by hywel on 5 Dec 2011 16:03 UTC**
I think this is fine, just a bit curious why the !DensityTrackerIndices enum has only one value assigned - presumably values should really be assigned for both enum possibilities?

Cheers

**Comment by schmie** [comment]
**Comment by miguel on 5 Dec 2011 16:08 UTC**
James integrated the incompressibility checker into the reporting infrastructure in 6c2a4f7cab68.

**Comment by schmie** [comment]
**Comment by miguel on 5 Dec 2011 16:16 UTC**
Replying to [hywel](comment:12):
> I think this is fine, just a bit curious why the !DensityTrackerIndices enum has only one value assigned - presumably values should really be assigned for both enum possibilities?

As far as I know, it will number the fields of the enum incrementally if you don't set a value explicitly.

**Comment by schmie** [comment]
**Comment by miguel on 9 Dec 2011 09:06 UTC**
Replying to [miguel](comment:11):
> To do:
>  * evaluate impact on performance of the checker.
>  * several TODO comments left in the code.

Added a new timer for code that monitors stability and incompressibility in 7d1421d77138. Todo comments marked with proper Doxygen (I think!).

James, please CR. If OK, we can move this ticket to 0.3. Performance evaluation should be probably done in a separate ticket. Only todo comments left and they are low priority.

**Comment by schmie** [comment]
**Modified by hywel on 13 Jan 2012 11:45 UTC**

**Comment by schmie** [comment]
**Comment by jamespjh on 13 Jan 2012 12:30 UTC**
I remember reviewing this verbally with Miguel at a computer. It's all good.

**Comment by schmie** [comment]
**Modified by hywel on 1 Mar 2012 17:47 UTC**

**Comment by schmie** [comment]
**Comment by jamespjh on 7 Mar 2012 20:24 UTC**
I reviewed at comment:17, this should be closed.

**Comment by schmie** [comment]
**Comment by hywel on 7 Mar 2012 20:30 UTC**
Milestone 0.3 deleted
